### PR TITLE
Add generate SBOM feature

### PR DESCRIPTION
**Pull request by stonezdj**
Thank you for contributing to Harbor!

# Comprehensive Summary of your change

# Issue being fixed
Fixes #(issue)

Please indicate you've done the following:
- [ ] Well Written Title and Summary of the PR
- [ ] Label the PR as needed. "release-note/ignore-for-release, release-note/new-feature, release-note/update, release-note/enhancement, release-note/community, release-note/breaking-change, release-note/docs, release-note/infra, release-note/deprecation"
- [ ] Accepted the DCO. Commits without the DCO will delay acceptance.
- [ ] Made sure tests are passing and test coverage is added if needed.
- [ ] Considered the docs impact and opened a new docs issue or PR with docs changes if needed in [website repository](https://github.com/goharbor/website).

### CODE REVIEW

File: src/controller/artifact/controller.go [hunk 1]
```diff
@@ -73,6 +74,8 @@ var (
 		chart.ArtifactTypeChart: icon.DigestOfIconChart,
 		cnab.ArtifactTypeCNAB:   icon.DigestOfIconCNAB,
 		wasm.ArtifactTypeWASM:   icon.DigestOfIconWASM,
+		// TODO: change it to the real icon of SBOM
```
Comment (wy65701436) [inline src/controller/artifact/controller.go:77]: please sync with the latest main code, which has the sbom icon support.

File: src/controller/scan/base_controller.go [hunk 2]
```diff
@@ -108,6 +110,7 @@ type basicController struct {
 	rc robot.Controller
 	// Tag controller
 	tagCtl tag.Controller
+	artCtl artifact.Controller
```
Comment (wy65701436) [inline src/controller/scan/base_controller.go:113]: comment?

File: src/controller/scanner/base_controller.go [hunk 3]
```diff
@@ -122,10 +126,34 @@ func (bc *basicController) GetRegistration(ctx context.Context, registrationUUID
...
 	return r, nil
 }
 
+func (bc *basicController) appendCap(ctx context.Context, r *scanner.Registration) error {
+	mt, err := bc.Ping(ctx, r)
+	if err != nil {
+		logger.Errorf("Get registration error: %s", err)
+		return err
+	}
+	capabilities := map[string]interface{}{}
```
Comment (wy65701436) [inline src/controller/scanner/base_controller.go:144]: can we define an spec object bases on the scanner spec?

File: src/common/rbac/const.go [hunk 4]
```diff
@@ -182,6 +183,9 @@ var (
 			{Resource: ResourceScan, Action: ActionRead},
 			{Resource: ResourceScan, Action: ActionStop},
 
+			{Resource: ResourceSBOM, Action: ActionCreate},
+			{Resource: ResourceSBOM, Action: ActionRead},
```
Comment (zyyw) [inline src/common/rbac/const.go:187]: also need `{Resource: ResourceSBOM, Action: ActionStop},`?

File: src/common/rbac/project/rbac_role.go [hunk 5]
```diff
@@ -86,6 +86,8 @@ var (
 			{Resource: rbac.ResourceScan, Action: rbac.ActionCreate},
 			{Resource: rbac.ResourceScan, Action: rbac.ActionRead},
 			{Resource: rbac.ResourceScan, Action: rbac.ActionStop},
+			{Resource: rbac.ResourceSBOM, Action: rbac.ActionCreate},
+			{Resource: rbac.ResourceSBOM, Action: rbac.ActionRead},
```
Comment (zyyw) [inline src/common/rbac/project/rbac_role.go:90]: also need `{Resource: ResourceSBOM, Action: ActionStop},`?

File: src/common/rbac/project/rbac_role.go [hunk 6]
```diff
@@ -169,6 +171,7 @@ var (
 			{Resource: rbac.ResourceScan, Action: rbac.ActionCreate},
 			{Resource: rbac.ResourceScan, Action: rbac.ActionRead},
 			{Resource: rbac.ResourceScan, Action: rbac.ActionStop},
+			{Resource: rbac.ResourceSBOM, Action: rbac.ActionRead},
```
Comment (zyyw) [inline src/common/rbac/project/rbac_role.go:174]: also need `{Resource: ResourceSBOM, Action: ActionStop},`?

File: src/controller/artifact/controller.go [hunk 7]
```diff
@@ -73,6 +74,7 @@ var (
 		chart.ArtifactTypeChart: icon.DigestOfIconChart,
 		cnab.ArtifactTypeCNAB:   icon.DigestOfIconCNAB,
 		wasm.ArtifactTypeWASM:   icon.DigestOfIconWASM,
+		sbom.ArtifactTypeSBOM: icon.DigestOfIconAccSBOM,
```
Comment (zyyw) [inline src/controller/artifact/controller.go:75]: gofmt? the `icon.DigestOfIconAccSBOM,` should be vertically align with the other values.

File: src/pkg/scan/sbom/sbom.go [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+// RequestProducesMineTypes defines the mine types produced by the scan handler
+func (v *scanHandler) RequestProducesMineTypes() []string {
+	return []string{v1.MimeTypeSBOMReport}
+}
+
+// RequestParameters defines the parameters for scan request
+func (v *scanHandler) RequestParameters() map[string]interface{} {
+	return map[string]interface{}{"sbom_media_types": []string{"application/spdx+json"}}
```
Comment (zyyw) [inline src/pkg/scan/sbom/sbom.go:59]: should we make `application/spdx+json` a constant variable? same comment for line 64.

File: api/v2.0/swagger.yaml [hunk 9]
```diff
@@ -1179,7 +1179,7 @@ paths:
         - name: scan_request_type
```
Comment (xuelichao) [inline api/v2.0/swagger.yaml:1179]: Shouldn't we rename the 'scan_request_type'. to 'scanType' to keep it same with stop API?

File: src/pkg/scan/sbom/sbom.go [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+		Registry: sr.Registry,
+		Artifact: sr.Artifact,
+	}
+	// the registry server url is core by default, need to replace it with real registry server url
+	scanReq.Registry.URL = v.RegistryServer(ctx.SystemContext())
+	if len(scanReq.Registry.URL) == 0 {
+		return "", fmt.Errorf("empty registry server")
+	}
+	myLogger := ctx.GetLogger()
+	myLogger.Debugf("Pushing accessory artifact to %s/%s", scanReq.Registry.URL, scanReq.Artifact.Repository)
```
Comment (wy65701436) [inline src/pkg/scan/sbom/sbom.go:101]: move the log into GenAccessoryFunc() if you like

File: src/pkg/scan/sbom/sbom.go [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+	summary[sbom.ScanStatus] = status
+	rep, err := json.Marshal(summary)
+	if err != nil {
+		return "", err
+	}
+	return string(rep), nil
+}
+
+// extract server name from config, and remove the protocol prefix
+func getRegistryServer(ctx context.Context) string {
```
Comment (wy65701436) [inline src/pkg/scan/sbom/sbom.go:127]: getRAWRegistryServer()

File: src/pkg/scan/job.go [hunk 12]
```diff
@@ -367,7 +373,19 @@ func ExtractScanReq(params job.Parameters) (*v1.ScanRequest, error) {
...
+	// attach the request with ProducesMimeTypes and Parameters
+	if len(req.RequestType) > 0 {
+		// current only support requestType with one element for each request
+		if len(req.RequestType[0].Type) > 0 {
+			reqType = req.RequestType[0].Type
+		}
+		handler := GetScanHandler(reqType)
+		if handler != nil {
+			req.RequestType[0].ProducesMimeTypes = handler.RequestProducesMineTypes()
+			req.RequestType[0].Parameters = handler.RequestParameters()
```
Comment (wy65701436) [inline src/pkg/scan/job.go:386]: please handle the nil situation.

File: src/pkg/scan/sbom/model/summary.go [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+	// Duration ...
+	Duration = "duration"
+	// ScanStatus ...
+	ScanStatus = "scan_status"
+)
+
+// Summary includes the sbom summary information
+type Summary map[string]interface{}
+
+// SBOMAccessory returns the repository and digest of the SBOM
```
Comment (wy65701436) [inline src/pkg/scan/sbom/model/summary.go:35]: SBOMAccessory => SBOMAccArt

File: src/go.sum [hunk 14]
```diff
@@ -703,6 +703,7 @@ golang.org/x/crypto v0.0.0-20210921155107-089bfa567519/go.mod h1:GvvjBRRGRdwPK5y
 golang.org/x/crypto v0.0.0-20211215153901-e495a2d5b3d3/go.mod h1:IxCIyHEi3zRg3s0A5j5BB6A9Jmi73HwBIUl50j+osU4=
 golang.org/x/crypto v0.0.0-20220622213112-05595931fe9d/go.mod h1:IxCIyHEi3zRg3s0A5j5BB6A9Jmi73HwBIUl50j+osU4=
 golang.org/x/crypto v0.13.0/go.mod h1:y6Z2r+Rw4iayiXXAIxJIDAJ1zMW4yaTpebo8fPOliYc=
+golang.org/x/crypto v0.19.0/go.mod h1:Iy9bg/ha4yyC70EfRS8jz+B6ybOBKMaSxLj6P6oBDfU=
```
Comment (wy65701436) [inline src/go.sum:706]: please revert these changes.

File: src/controller/scan/base_controller.go [hunk 15]
```diff
@@ -1265,3 +1275,45 @@ func parseOptions(options ...Option) (*Options, error) {
...
+	}
+	return nil
+}
+
+// deleteArtifactAccessory check if current report has accessory info, if there is, delete it
+func (bc *basicController) deleteArtifactAccessory(ctx context.Context, report string) error {
+	if len(report) == 0 {
+		return nil
+	}
+	sbomSummary := sbomModel.Summary{}
```
Comment (wy65701436) [inline src/controller/scan/base_controller.go:1294]: please use the mimetype to judge the report type.

File: src/pkg/scan/dao/scanner/model.go [hunk 16]
```diff
@@ -151,7 +152,11 @@ func (r *Registration) HasCapability(manifestMimeType string) bool {
 }
 
 // GetProducesMimeTypes returns produces mime types for the artifact
-func (r *Registration) GetProducesMimeTypes(mimeType string) []string {
+func (r *Registration) GetProducesMimeTypes(mimeType string, scanType string) []string {
+	// specify the mime types returned in the request
```
Comment (wy65701436) [inline src/pkg/scan/dao/scanner/model.go:156]: I suppose this is the code for prototype, please leverage the type in the capability to revise this.

File: api/v2.0/swagger.yaml [hunk 17]
```diff
@@ -6769,7 +6769,7 @@ definitions:
   ScanRequestType:
```
Comment (wy65701436) [inline api/v2.0/swagger.yaml:6769]: TODO: merge the ScanRequestType and ScanType

File: src/controller/scanner/base_controller.go [hunk 18]
```diff
@@ -122,10 +126,25 @@ func (bc *basicController) GetRegistration(ctx context.Context, registrationUUID
...
+	if r == nil {
+		return nil, nil
+	}
+	if err := bc.appendCap(ctx, r); err != nil {
+		return nil, err
+	}
 	return r, nil
 }
 
+func (bc *basicController) appendCap(ctx context.Context, r *scanner.Registration) error {
```
Comment (wy65701436) [inline src/controller/scanner/base_controller.go:138]: just use the capability() as the func name.

File: src/controller/scan/base_controller.go (hunk 2 above)
Comment (stonezdj) [reply src/controller/scan/base_controller.go:113]: Updated

File: src/pkg/scan/sbom/sbom.go (hunk 11 above)
Comment (stonezdj) [reply src/pkg/scan/sbom/sbom.go:138]: registryFQDN()

**Review by wy65701436 — APPROVED** [review]
lgtm

**Review by zyyw — APPROVED** [review]
